### PR TITLE
✨ Add Cronet embedded tool

### LINKED ISSUE #807: cronet_http: depend on org.chromium.net:cronet-api instead
The current implementation depends on `com.google.android.gms:play-services-cronet`, which pulls in the Google Play services. This might be undesirable for apps that do not want to have GMS code in them (and those who mainly target Chinese users and do not have GMS available).

We should depend on Chromium's API package, which will find the relevant dependencies automatically if they are declared. Additional documentations should be provided to the users on how to choose a Cronet provider (similar to https://exoplayer.dev/network-stacks.html#cronet).

### PR DESCRIPTION
**Pull request by AlexV525**
Resolves #807.

Creates a new tool for publishing the non-GMS-based Cronet package. The tool is implemented with these concerns:
* Not creating symlinks because the structure can be changed easily.
* Not to restrict running on a single platform, so it's been written in Dart.
* Only modify files, but do not publish in this tool, leave the process to the operator, since the latest stable version can have breaking changes and require further CI tests (more jobs in `.github/workflows`).
* As easy as possible to maintain the tool.

### CODE REVIEW

**Comment by AlexV525** [comment]
@brianquinlan @natebosch

File: pkgs/cronet_http/tool/prepare_for_embedded.dart [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    'dl.google.com',
+    'android/maven2/org/chromium/net/group-index.xml',
+  );
+  final response = await http.get(url);
+  final parsedXml = XmlDocument.parse(response.body);
+  final embeddedNode = parsedXml.children
+      .singleWhere((e) => e is XmlElement)
+      .children
+      .singleWhere((e) => e is XmlElement && e.name.local == 'cronet-embedded');
+  final stableVersionReg = RegExp(r'^(\d+).(\d+).(\d+)$');
```
Comment (AlexV525) [inline pkgs/cronet_http/tool/prepare_for_embedded.dart:26]:
```suggestion
  final stableVersionReg = RegExp(r'^\d+.\d+.\d+$');
```

File: pkgs/cronet_http/tool/prepare_for_embedded.dart [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+void _writeImplementationToTheFile(String latestVersion) {
+  var dir = Directory.current;
+  if (dir.path.endsWith('tool')) {
+    dir = dir.parent;
+  }
+  // Update android/build.gradle
+  final fBuildGradle = File('${dir.path}/android/build.gradle');
+  final gradleContent = fBuildGradle.readAsStringSync();
+  final implementationRegExp = RegExp(
+    '^(\\s*)implementation [\'"]'
```
Comment (AlexV525) [inline pkgs/cronet_http/tool/prepare_for_embedded.dart:44]:
```suggestion
    '^\\s*implementation [\'"]'
```

File: pkgs/cronet_http/tool/prepare_for_embedded.dart [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    'com.google.android.gms:play-services-cronet'
+    ':(\\d+.\\d+.\\d+)[\'"]',
```
Comment (AlexV525) [inline pkgs/cronet_http/tool/prepare_for_embedded.dart:46]:
```suggestion
    ':\\d+.\\d+.\\d+[\'"]',
```

**Comment by brianquinlan** [comment]
@natebosch Do you know how the upcoming auto-publish feature works? Will it be possible to trigger a script like this as part of the publishing process?

**Comment by AlexV525** [comment]
FYI, previously when we do automatically publish on GitHub, it's easy to run any scripts in jobs. e.g. https://github.com/fluttercandies/flutter_photo_manager/blob/361773c98c4f9665ee61ef25f1a6aa003d7692ec/.github/workflows/publish.yml#L14

**Comment by natebosch** [comment]
@devoncarew is setting up autopublish on some repos. I don't think we've set up any mono repos yet - so we may need some work in the `mono_repo` tool.

**Comment by devoncarew** [comment]
From my understanding, you want to run a custom dart script before publishing? It'll perform a native build or provision native artifacts?

The new CI publishing feature has a few different ways it can operate. The way I've used it is configure the package on pub.dev to be published from a github action. Then, when a github action is triggered by a tag on a commit, running `dart pub publish --force` will recognize that it's running in a github action, has an 'OIDC' token, and pass that as authentication to pub.dev when publishing.

The tooling I've written for publishing automation - that does validation for PRs and publishing for tagged commits on main - is likely too purpose built to re-use here (it does however support mono-repos :) ).

If you think you want to use some publishing automation here, you could:
- add a `.github/workflows/publish_cronet_http.yaml` file, loosely based on https://github.com/devoncarew/firehose/blob/main/.github/workflows/publish_internal.yaml
- use the pattern `cronet_http-v[0-9]+.[0-9]+.[0-9]+*` as the tag's glob pattern to trigger on
- in that workflow file, you want to make sure to have the right cwd for your job steps (`working-directory: foo`)
- run whatever arbitrary script you want
- finish with a call to `dart pub publish --force`

**Comment by AlexV525** [comment]
Okay according to the above discussions and advises from @devoncarew, we can make an auto-publish workflow, and also allow it to be done manually using the simple Dart tool. I'll go for this goal then, please vote 👍 if you guys think it's valid.

**Comment by brianquinlan** [comment]
@AlexV525 Your solution will work but it feels a bit hacky to me. 

Let's take a step make and think about options:
1. Fork this repo, make the needed changes, and periodically sync the fork:
    - The sync can be [automated](https://github.com/marketplace/actions/fork-sync)
    - There would be no coordination in releases i.e. a release of `package:cronet_http` would not automatically release `package:cronet_http_embedded`
    -  The fork would include unnecessary packages like `cupertino_http`
2. Create a script that modifies this repository in-place (this approach): 
    - Error prone if used done manually (but reasonable if done through automation)
    - Will double CI time if also runs with the script-modified changes
3. Create a branch, make the needed changes, and periodically sync from main
    -  Similar to (1) except the Dart team would have to manage it

Is there some other approach?

**Comment by AlexV525** [comment]
The main guideline I based on when I was developing ideas is **Avoid manipulation by humans as possible, let the automation rolls**. The guideline will avoid exceptions made by human mistakes, such as forgetting to apply patches, forgetting to sync the code base, merging leads to code loss, etc.

I'll base on this guideline to discuss the above options.

### Fork this repo, make the needed changes, and periodically sync the fork

>    * The sync can be [automated](https://github.com/marketplace/actions/fork-sync)
>    * There would be no coordination in releases i.e. a release of `package:cronet_http` would not automatically release `package:cronet_http_embedded`

Not only the sync can be done automatically, as we discussed, but the publishing can also do it as well. The coordination can be bridged using GitHub Actions too, but it'll involve crossing repo calls, which seems too much for the proposal.

>    * The fork would include unnecessary packages like `cupertino_http`

That is correct, but the impact is not as much as we thought.

### Create a script that modifies this repository in-place (this approach)

>    * Error prone if used done manually (but reasonable if done through automation)

Given the complexity and the corresponding change made by the tool, I'll say it'll barely meet exceptions during regular development. The idea is to get rid of manual publishing, so if we think too much about doing this manually, we'll come to another path eventually.

>    * Will double CI time if also runs with the script-modified changes

The time cost can be replaced with space cost (running two jobs parallel). And if new exceptions were thrown during the publishing of the embedded package, it can be easily solved during PRs. It actually saves time for debugging incompatible changes.

### Create a branch, make the needed changes, and periodically sync from main

>    * Similar to (1) except the Dart team would have to manage it

Any manual manipulation should be avoided when doing patches or publishing, which will make the package solid.

### Other approaches?

I don't come up with any at present.

**Comment by AlexV525** [comment]
Frankly, the difference between the GMS-based and embedded is tiny in the code base, and the community as well as [the Dart team has already explored automatic publishing](https://dart.dev/tools/pub/automated-publishing). We should not have many issues with maintaining a publishing tool.

**Comment by AlexV525** [comment]
Hi @brianquinlan, did you come up with more thoughts or considerations in these weeks?

**Comment by brianquinlan** [comment]
Hey @AlexV525 ,

I'm patching in the code now to take a look.

**Comment by brianquinlan** [comment]
Hey @AlexV525, 

I think that you need to make at least two changes:
1. you need to modify the `name` and `description` in `pubspec.yaml`. The name should probably be something like `cronet_http_embedded` and the `description` could be `An Android Flutter plugin that provides access to the Cronet HTTP client. Identical to package:cronet_http except that it embeds Cronet rather than relying on Google Play Services.
2. we probably need a new readme (maybe `README_EMBEDDED.md`) that explains that the difference between it and `package:cronet_http` is and then links to `package:cronet_http` (so we don't need to keep the documentation in sync).
3. Your `prepare_for_embedded.dart` script should rename `README_EMBEDDED.md` to `README.md`

File: pkgs/cronet_http/pubspec.yaml [hunk 4]
```diff
@@ -16,6 +16,7 @@ dependencies:
 dev_dependencies:
   lints: ^1.0.0
   pigeon: ^3.2.3
+  xml: ^6.1.0
```
Comment (kevmoo) [inline pkgs/cronet_http/pubspec.yaml:19]: I got freaked out. It's only a dev_dependency. 😌

**Comment by AlexV525** [comment]
@brianquinlan Updated accroding to your suggestions.

**Comment by AlexV525** [comment]
Also please let me know if we can setup a GitHub action for this, maybe in separate PRs.

File: pkgs/cronet_http/README_EMBEDDED.md [hunk 5]
```diff
@@ -0,0 +1,23 @@
+An Android Flutter plugin that provides access to the
+[Cronet](https://developer.android.com/guide/topics/connectivity/cronet/reference/org/chromium/net/package-summary)
+HTTP client.
+
+Cronet embedded is available as a standalone implementation
+which will brought additional 8MB for apps approximately.
+
+## Status: Experimental
```
Comment (brianquinlan) [inline pkgs/cronet_http/README_EMBEDDED.md:8]: I'd remove this line and everything after it. It will be annoying to keep this file in sync with README.md

File: pkgs/cronet_http/tool/prepare_for_embedded.dart [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:io';
+
+import 'package:http/http.dart' as http;
+import 'package:xml/xml.dart';
+import 'package:yaml_edit/yaml_edit.dart';
+
+void main() async {
+  final latestVersion = await _getLatestVersion();
```
Comment (brianquinlan) [inline pkgs/cronet_http/tool/prepare_for_embedded.dart:12]: I have a few code nits.

I'd create three functions that do the three things that you want to do here:
1. `updateCronetDependency`
2. `updatePubSpec`
3. `updateReadme`

You can keep `_getLatestVersion` as a helper but I'd call it `_getLatestCronetVersion`.

File: pkgs/cronet_http/tool/prepare_for_embedded.dart [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+  final fBuildGradle = File('${dir.path}/android/build.gradle');
+  final gradleContent = fBuildGradle.readAsStringSync();
+  final implementationRegExp = RegExp(
+    '^\\s*implementation [\'"]'
+    'com.google.android.gms:play-services-cronet'
+    ':\\d+.\\d+.\\d+[\'"]',
+    multiLine: true,
+  );
+  final newImplementation = 'org.chromium.net:cronet-embedded:$latestVersion';
+  print('Patching $newImplementation');
```
Comment (brianquinlan) [inline pkgs/cronet_http/tool/prepare_for_embedded.dart:52]: You can remove this.

File: pkgs/cronet_http/tool/prepare_for_embedded.dart [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    '    implementation $newImplementation',
+  );
+  fBuildGradle.writeAsStringSync(newGradleContent);
+  // Update pubspec.yaml
+  final fPubspec = File('${dir.path}/pubspec.yaml');
+  final yamlEditor = YamlEditor(fPubspec.readAsStringSync())
+    ..update(['name'], 'cronet_http_embedded')
+    ..update(
+      ['description'],
+      'An Android Flutter plugin that '
```
Comment (brianquinlan) [inline pkgs/cronet_http/tool/prepare_for_embedded.dart:64]: Maybe make this a global constant?

File: pkgs/cronet_http/tool/prepare_for_embedded.dart [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+  print('Patching $newImplementation');
+  final newGradleContent = gradleContent.replaceAll(
+    implementationRegExp,
+    '    implementation $newImplementation',
+  );
+  fBuildGradle.writeAsStringSync(newGradleContent);
+  // Update pubspec.yaml
+  final fPubspec = File('${dir.path}/pubspec.yaml');
+  final yamlEditor = YamlEditor(fPubspec.readAsStringSync())
+    ..update(['name'], 'cronet_http_embedded')
```
Comment (brianquinlan) [inline pkgs/cronet_http/tool/prepare_for_embedded.dart:61]: Maybe make this a global constant?

File: pkgs/cronet_http/tool/run_pigeon.sh [hunk 10]
```diff
@@ -1,6 +1,7 @@
 #!/bin/sh
 
 # Generate the platform messages used by cronet_http.
+cd ../
```
Comment (brianquinlan) [inline pkgs/cronet_http/tool/run_pigeon.sh:4]: Why did you change this?

**Comment by brianquinlan** [comment]
I have a few code nits but overall it looks good!

I wouldn't worry about having a github action for now. I think that we are going to roll out automatic publishing soon and we can use this script when that rolls out. For now I'll just use it manually when I publish package:cronet_http.

File: pkgs/cronet_http/tool/prepare_for_embedded.dart [hunk 11]
```diff
@@ -0,0 +1,79 @@
+// Copyright (c) 2022, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
```
Comment (brianquinlan) [inline pkgs/cronet_http/tool/prepare_for_embedded.dart:4]: I think that we need a big comment here to explain what this does. Like:

/// The `cronet_http` directory is used to produce two packages:
/// 1. `package:cronet_http`, which uses the Google Play Services version of Cronet
/// 2. `package:cronet_http_embedded`, which embeds Cronet
///
/// The default configuration of this code is to use the Google Play Services version of Cronet.
///
/// This script transforms the default configuration into one that embeds Cronet by:
/// 1. modifying the Gradle build file to reference the embedded version of Cronet
/// 2. modifying the `name` and `description` in `pubspec.yaml`
/// 3. replaying `README.md` with `README_EMBEDDED.md`
///
/// After running this script, `flutter pub publish` can be run to update `package:cronet_http_embedded`.
///
/// **NOTE:*** This script modifies the above files **in place**.

**Review by brianquinlan — CHANGES_REQUESTED** [review]

File: pkgs/cronet_http/tool/prepare_for_embedded.dart (hunk 7 above)
Comment (AlexV525) [reply pkgs/cronet_http/tool/prepare_for_embedded.dart:52]: I thought this would be helpful to let you know which version of the embedded cronet you're using. WDYT?

File: pkgs/cronet_http/README_EMBEDDED.md (hunk 5 above)
Comment (AlexV525) [reply pkgs/cronet_http/README_EMBEDDED.md:8]: I think we can leave at least a *more details* link like:

> See more details about the cronet_http at: https://pub.dev/packages/cronet_http.

File: pkgs/cronet_http/tool/run_pigeon.sh (hunk 10 above)
Comment (AlexV525) [reply pkgs/cronet_http/tool/run_pigeon.sh:4]: It's under the root path previously, now it's under `/tool`.

File: pkgs/cronet_http/README_EMBEDDED.md [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+HTTP client.
+
+This package is identical to [`package:cronet_http`](https://pub.dev/packages/cronet_http)
+except that it embeds
+[Cronet](https://developer.android.com/guide/topics/connectivity/cronet/reference/org/chromium/net/package-summary)
+rather than using the version included with
+[Google Play Services](https://developers.google.com/android/guides/overview).
+This increases the uncompressed size of the application by approximately 8MB.
+
+See more details about the cronet_http at: https://pub.dev/packages/cronet_http.
```
Comment (brianquinlan) [inline pkgs/cronet_http/README_EMBEDDED.md:12]: Remove "the"

File: pkgs/cronet_http/tool/prepare_for_embedded.dart [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+// BSD-style license that can be found in the LICENSE file.
+// ignore_for_file: lines_longer_than_80_chars
+
+import 'dart:io';
+
+import 'package:http/http.dart' as http;
+import 'package:xml/xml.dart';
+import 'package:yaml_edit/yaml_edit.dart';
+
+/// The cronet_http directory is used to produce two packages:
```
Comment (brianquinlan) [inline pkgs/cronet_http/tool/prepare_for_embedded.dart:12]: I think that convention is that you put the file comment before the imports.

File: pkgs/cronet_http/tool/prepare_for_embedded.dart [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+/// - `cronet_http`, which uses the Google Play Services version of Cronet.
+/// - `cronet_http_embedded`, which embeds Cronet.
+///
+/// The default configuration of this code is to use the Google Play Services version of Cronet.
+///
+/// The script transforms the configuration into one that embeds Cronet by:
+/// 1. Modifying the Gradle build file to reference the embedded version of Cronet.
+/// 2. Modifying the *name* and *description* in `pubspec.yaml`.
+/// 3. Replaying `README.md` with `README_EMBEDDED.md`.
```
Comment (brianquinlan) [inline pkgs/cronet_http/tool/prepare_for_embedded.dart:21]: Replacing

File: pkgs/cronet_http/tool/prepare_for_embedded.dart [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+///
+/// After running this script, `flutter pub publish` can be run to update package:cronet_http_embedded.
```
Comment (brianquinlan) [inline pkgs/cronet_http/tool/prepare_for_embedded.dart:23]: Wrap at 80 columns please.

File: pkgs/cronet_http/tool/prepare_for_embedded.dart [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+///
+/// NOTE: This script modifies the above files in place.
+late final Directory _directory;
+
+const String _gmsDependencyName = 'com.google.android.gms:play-services-cronet';
```
Comment (brianquinlan) [inline pkgs/cronet_http/tool/prepare_for_embedded.dart:28]: All of these type declarations are redundant, please remove them i.e. just write
const _gmsDependencyName = "...";

File: pkgs/cronet_http/tool/prepare_for_embedded.dart [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+///
+/// The script transforms the configuration into one that embeds Cronet by:
+/// 1. Modifying the Gradle build file to reference the embedded version of Cronet.
+/// 2. Modifying the *name* and *description* in `pubspec.yaml`.
+/// 3. Replaying `README.md` with `README_EMBEDDED.md`.
+///
+/// After running this script, `flutter pub publish` can be run to update package:cronet_http_embedded.
+///
+/// NOTE: This script modifies the above files in place.
+late final Directory _directory;
```
Comment (brianquinlan) [inline pkgs/cronet_http/tool/prepare_for_embedded.dart:26]: Maybe `_packageDirectory`?

File: pkgs/cronet_http/tool/prepare_for_embedded.dart (hunk 7 above)
Comment (brianquinlan) [reply pkgs/cronet_http/tool/prepare_for_embedded.dart:80]: OK.

File: pkgs/cronet_http/tool/run_pigeon.sh (hunk 10 above)
Comment (brianquinlan) [reply pkgs/cronet_http/tool/run_pigeon.sh:4]: OK.

**Review by brianquinlan — CHANGES_REQUESTED** [review]

File: pkgs/cronet_http/tool/prepare_for_embedded.dart (hunk 16 above)
Comment (AlexV525) [reply pkgs/cronet_http/tool/prepare_for_embedded.dart:28]: Used to write with Flutter. :)

**Comment by brianquinlan** [comment]
Hey @AlexV525,

Thanks for your contribution and your patience!

I will try to do a release tomorrow.

**Review by brianquinlan — APPROVED** [review]

**Comment by AlexV525** [comment]
> I will try to do a release tomorrow.

Is it working as expected?

**Comment by brianquinlan** [comment]
@AlexV525 The release is https://pub.dev/packages/cronet_http_embedded

You could check to see if it works as you expected ;-)

**Comment by DemoJameson** [comment]
build failed with cronet_http_embedded on android, since `implementation com.google.android.gms:play-services-cronet:18.0.1` lack of double quotes
```
FAILURE: Build failed with an exception.

* Where:
Build file 'C:\Users\DemoJameson\AppData\Local\Pub\Cache\hosted\pub.flutter-io.cn\cronet_http_embedded-0.2.0\android\build.gradle' line: 49

* What went wrong:
Could not compile build file 'C:\Users\DemoJameson\AppData\Local\Pub\Cache\hosted\pub.flutter-io.cn\cronet_http_embedded-0.2.0\android\build.gradle'.
> startup failed:
  build file 'C:\Users\DemoJameson\AppData\Local\Pub\Cache\hosted\pub.flutter-io.cn\cronet_http_embedded-0.2.0\android\build.gradle': 49: Unexpected input: '{' @ line 49, column 14.
     dependencies {
                  ^

  1 error
```

**Comment by brianquinlan** [comment]
I just released version 0.2.1 - could you check to see if that fixes the problem?
